File: android/src/main/java/io/parity/signer/components/networkicon/jdenticon/jdenticon_kotlin/IconGenerator.kt
package io.parity.signer.components.networkicon.jdenticon.jdenticon_kotlin

import kotlin.math.floor

internal class IconGenerator(renderer: SvgRenderer, hash: String,
														 x: Float, y: Float, size: Float,
														 padding: Float?, config: Config,
	) {

    var _hash = hash
    val _renderer = renderer

    // Calculate padding
    var _padding = floor(size * (padding ?: 0f))
    var _size = size - _padding * 2f

    var graphics = Graphics(renderer)

    // Calculate cell size and ensure it is an integer
    var cell = floor(_size / 4f)

    // Since the cell size is integer based, the actual icon will be slightly smaller than specified => center icon
    var _x = x + floor(_padding + _size / 2f - cell * 2f);
    var _y = y + floor(_padding + _size / 2f - cell * 2f);

    // AVAILABLE COLORS
    var hue = hash.substring(hash.length-7).toInt(16).toFloat() / 0xfffffff

    // Available colors for this icon
    var availableColors = colorTheme(hue.toFloat(), config)

    // The index of the selected colors
    var selectedColorIndexes = ArrayList<Int>()
    var index = 0

    fun renderShape(
			colorIndex: Int,
			shapes: List<(Graphics, Float, Int?) -> Unit>,
			index: Int,
			rotationIndex: Int?,
			positions: Array<Array<Int>>
    ) {
        var r = if(rotationIndex != null)_hash.elementAt(rotationIndex).toString().toInt(16) else 0
        val shape = shapes[_hash.elementAt(index).toString().toInt(16) % shapes.size]

        _renderer.beginShape(availableColors[selectedColorIndexes[colorIndex]])
        for (i in 0 until positions.size) {
            graphics._transform = Transform(
                    _x + positions[i][0] * cell,
                    _y + positions[i][1] * cell,
                    cell,
                    ((r++ % 4).toFloat())
            )
            shape(graphics, cell, i)
        }

        _renderer.endShape()
    }

    fun isDuplicate(values: List<Int>) : Boolean {
        if (values.indexOf(index) >= 0) {
            for (i in 0 until values.size) {
                if (selectedColorIndexes.indexOf(values[i]) >= 0) {
                    return true
                }
            }
        }
        return false
    }

    init {
        for (i in 0 until 3) {
            index = (hash.elementAt(8 + i).toString().toInt(16) % availableColors.size)
            if (isDuplicate(listOf(0, 4)) || // Disallow dark gray and dark color combo
                    isDuplicate(listOf(2, 3))) { // Disallow light gray and light color combo
                index = 1;
            }
            selectedColorIndexes.add(index);
        }
        // ACTUAL RENDERING
        // Sides
        renderShape(0,
					Shapes.outer, 2, 3, arrayOf(arrayOf(1, 0), arrayOf(2, 0), arrayOf(2, 3), arrayOf(1, 3), arrayOf(0, 1), arrayOf(3, 1), arrayOf(3, 2), arrayOf(0, 2)))
        // Corners
        renderShape(1,
					Shapes.outer, 4, 5, arrayOf(arrayOf(0, 0), arrayOf(3, 0), arrayOf(3, 3), arrayOf(0, 3)))
        // Center
        renderShape(2,
					Shapes.center, 1, null, arrayOf(arrayOf(1, 1), arrayOf(2, 1), arrayOf(2, 2), arrayOf(1, 2)))

        _renderer.finish()
    }
};

File: ios/Packages/Jdenticon/Sources/Jdenticon/Drawing/IconGenerator.swift
//
//  IconGenerator.swift
//
//
//  Created by Krzysztof Rodak on 22/08/2023.
//

import CoreGraphics
import Foundation

class IconGenerator {
    private var _hash: String
    private let _renderer: Renderer

    private var _padding: CGFloat
    private var _size: CGFloat
    private var graphics: Graphics

    private var cell: CGFloat
    private var _x: CGFloat
    private var _y: CGFloat

    private var hue: CGFloat
    private var availableColors: [String]

    private var selectedColorIndexes: [Int] = []
    private var index: Int = 0

    init(
        renderer: Renderer,
        hash: String,
        x: CGFloat,
        y: CGFloat,
        size: CGFloat,
        padding: CGFloat?,
        config: Config
    ) {
        _hash = hash
        _renderer = renderer
        _x = x
        _y = y
        _padding = floor(size * (padding ?? 0))
        _size = size - _padding * 2.0

        graphics = Graphics(renderer: renderer)

        cell = floor(_size / 4.0)

        _x += floor(_padding + _size / 2.0 - cell * 2.0)
        _y += floor(_padding + _size / 2.0 - cell * 2.0)

        let hueHex = String(hash.suffix(7))
        let hueInt = UInt32(hueHex, radix: 16) ?? 0
        hue = CGFloat(hueInt) / 0xFFFFFFF

        availableColors = colorTheme(hue: hue, config: config)

        for i in 0 ..< 3 {
            let charIndex = hash.index(hash.startIndex, offsetBy: 8 + i)
            let char = hash[charIndex]
            if let intValue = Int(String(char), radix: 16) {
                index = intValue % availableColors.count
                if isDuplicate(values: [0, 4]) || isDuplicate(values: [2, 3]) {
                    index = 1
                }
                selectedColorIndexes.append(index)
            }
        }

        renderShapes()
        _renderer.finish()
    }

    func renderShapes() {
        // Sides
        renderShape(
            colorIndex: 0,
            shapes: Shapes.outer,
            index: 2,
            rotationIndex: 3,
            positions: [
                [1, 0], [2, 0], [2, 3], [1, 3],
                [0, 1], [3, 1], [3, 2], [0, 2]
            ]
        )

        // Corners
        renderShape(
            colorIndex: 1,
            shapes: Shapes.outer,
            index: 4,
            rotationIndex: 5,
            positions: [
                [0, 0], [3, 0], [3, 3], [0, 3]
            ]
        )

        // Center
        renderShape(
            colorIndex: 2,
            shapes: Shapes.center,
            index: 1,
            rotationIndex: nil,
            positions: [
                [1, 1], [2, 1], [2, 2], [1, 2]
            ]
        )
    }

    func renderShape(
        colorIndex: Int,
        shapes: [(Graphics, CGFloat, Int?) -> Void],
        index: Int,
        rotationIndex: Int?,
        positions: [[Int]]
    ) {
        var r: Int = if let rotationIndex {
            Int(String(_hash[_hash.index(_hash.startIndex, offsetBy: rotationIndex)]), radix: 16) ?? 0
        } else {
            0
        }

        let shape =
            shapes[(Int(String(_hash[_hash.index(_hash.startIndex, offsetBy: index)]), radix: 16) ?? 0) % shapes
                .count]

        let color = availableColors[selectedColorIndexes[colorIndex]]
        _renderer.beginShape(color: color)

        for i in 0 ..< positions.count {
            graphics.transform = Transform(
                x_t: _x + CGFloat(positions[i][0]) * cell,
                y_t: _y + CGFloat(positions[i][1]) * cell,
                size_t: cell,
                rotation_t: CGFloat(r % 4)
            )
            shape(graphics, cell, i)
            r += 1
        }

        _renderer.endShape()
    }

    func isDuplicate(values: [Int]) -> Bool {
        if values.firstIndex(of: index) != nil {
            for i in 0 ..< values.count {
                if selectedColorIndexes.contains(values[i]) {
                    return true
                }
            }
        }
        return false
    }
}
